voyager-publish/index.js

```diff
@@ -9,28 +9,60 @@ export const publish = async ({
   maxMeasurements = 1000,
   logger = console
 }) => {
+  // Lock measurements for this run
+  // - On success, measurements will be deleted
+  // - On failure, lock will be released after 5 minutes
+  // TODO: Ensure `publishStart` is unique
+  const publishStart = new Date()
+
   // Fetch measurements
-  const { rows: measurements } = await pgPool.query(`
-    SELECT
-      id,
-      zinnia_version,
-      participant_address,
-      status_code,
-      end_at,
-      inet_group,
-      car_too_large,
-      cid
-    FROM measurements
-    LIMIT $1
-  `, [
-    maxMeasurements
-  ])
+  let measurements
+  {
+    const pgClient = await pgPool.connect()
+    try {
+      await pgClient.query('BEGIN ISOLATION LEVEL SERIALIZABLE')
+
+      // Delete published measurements
+      const { rows } = await pgClient.query(`
+        WITH rows AS (
+          SELECT id
+          FROM measurements
+          WHERE publish_start IS NULL OR publish_start < NOW() - INTERVAL '5 minutes'
+          ORDER BY id
+          LIMIT $1
+        )
+        UPDATE measurements
+        SET publish_start = $2
+        WHERE EXISTS (SELECT * FROM rows WHERE measurements.id = rows.id)
+        RETURNING
+          id,
+          zinnia_version,
+          participant_address,
+          status_code,
+          end_at,
+          inet_group,
+          car_too_large,
+          cid
+      `, [
+        maxMeasurements,
+        publishStart
+      ])
+      measurements = rows
+
+      await pgClient.query('COMMIT')
+    } catch (err) {
+      await pgClient.query('ROLLBACK')
+      throw err
+    } finally {
+      pgClient.release()
+    }
+  }
 
   // Fetch the count of all unpublished measurements - we need this for monitoring
   // Note: this number will be higher than `measurements.length` because voyager-api adds more
   // measurements in between the previous and the next query.
   const totalCount = (await pgPool.query(
-    'SELECT COUNT(*) FROM measurements'
+    'SELECT COUNT(*) FROM measurements WHERE publish_start IS NULL'
   )).rows[0].count
 
   logger.log(`Publishing ${measurements.length} measurements. Total unpublished: ${totalCount}. Batch size: ${maxMeasurements}.`)
```

Comment:
(1)
This is changing the behaviour of this code. So far, we have been counting all measurements waiting to be published so that we can get alerted if our publishing pipeline is not able to keep up with the inflow.

I think your new version makes this behaviour subtly different. Together with the 5-minute lock timeout, it will delay the alert trigger by that 5 minutes. There could be more differences that I am not able to foresee now.

What is your reasoning for adding the new WHERE condition?

(2)
We are going to increase the frequency of how often we report this number. I guess it's not a big deal now. It makes me wonder, though, if it would be better to move this part from the worker up to the parent process.

Reply:
(1) Good point! You're right, I have reverted the query change
(2) Let's fix it when it becomes a problem?